File: dataflow/scheduler.py
"""Thread-per-node Scheduler with metrics collection."""

import threading
import time
from dataclasses import dataclass, field

from dataflow.primitives import _STOP
from dataflow.node import Node
from dataflow.graph import Graph


@dataclass
class NodeMetrics:
    """Per-node execution metrics."""
    name: str
    items_processed: int = 0
    elapsed: float = 0.0
    throughput: float = 0.0
    error: str | None = None


class Scheduler:
    """Runs each node in its own thread, in topological order."""

    def __init__(self, graph: Graph):
        self.graph = graph
        self.wall_time: float = 0.0
        self._metrics: dict[str, NodeMetrics] = {}

    def run(self) -> dict[str, NodeMetrics]:
        """Start all node threads, join them, and return metrics."""
        self.graph.validate()
        order = self.graph.topological_order()

        threads: list[threading.Thread] = []
        for node in order:
            t = threading.Thread(
                target=self._run_node,
                args=(node,),
                name=f"node-{node.name}",
                daemon=True,
            )
            threads.append(t)

        t0 = time.perf_counter()
        for t in threads:
            t.start()
        for t in threads:
            t.join()
        self.wall_time = time.perf_counter() - t0

        return dict(self._metrics)

    def _run_node(self, node: Node):
        """Thread target: run node, catch exceptions, propagate _STOP on error."""
        metrics = NodeMetrics(name=node.name)
        try:
            node.run()
            metrics.items_processed = node.items_processed
            metrics.elapsed = node.elapsed
            if metrics.elapsed > 0:
                metrics.throughput = metrics.items_processed / metrics.elapsed
        except Exception as e:
            metrics.error = str(e)
            # Propagate _STOP downstream so dependent nodes shut down
            for port in node.outputs.values():
                if port.connected:
                    try:
                        port.put(_STOP)
                    except Exception:
                        pass
        finally:
            self._metrics[node.name] = metrics

    @staticmethod
    def print_metrics(metrics: dict[str, NodeMetrics], wall_time: float = 0.0):
        """Print a formatted metrics table."""
        print(f"\n{'Node':<20} {'Items':>10} {'Time (s)':>10} {'Throughput':>14} {'Error'}")
        print("-" * 70)
        for m in metrics.values():
            thr = f"{m.throughput:>12.0f}/s" if m.throughput > 0 else "           -"
            err = m.error or ""
            print(f"{m.name:<20} {m.items_processed:>10} {m.elapsed:>10.4f} {thr} {err}")
        if wall_time > 0:
            print(f"\nWall time: {wall_time:.4f}s")

File: graphos-cpp/src/dataflow/scheduler.cpp
#include "graphos/dataflow/scheduler.hpp"
#include <chrono>
#include <iostream>
#include <iomanip>
#include <mutex>

namespace graphos {

std::unordered_map<std::string, NodeMetrics> Scheduler::run() {
    graph_.validate();
    auto order = graph_.topological_order();

    std::unordered_map<std::string, NodeMetrics> metrics;
    std::mutex metrics_mu;

    auto t0 = std::chrono::steady_clock::now();

    // Launch one jthread per node in topological order
    threads_.reserve(order.size());
    for (auto& node_ptr : order) {
        threads_.emplace_back([&node_ptr, &metrics, &metrics_mu, this](
                                  std::stop_token st) {
            NodeMetrics m;
            m.name = node_ptr->name();
            try {
                node_ptr->run(st);
                m.items_processed = node_ptr->items_processed();
                m.elapsed_s = node_ptr->elapsed();
                if (m.elapsed_s > 0.0)
                    m.throughput = static_cast<double>(m.items_processed) /
                                  m.elapsed_s;
            } catch (const std::exception& e) {
                m.error = e.what();
                m.items_processed = node_ptr->items_processed();
                m.elapsed_s = node_ptr->elapsed();
            } catch (...) {
                m.error = "unknown exception";
            }
            std::lock_guard lock(metrics_mu);
            metrics[m.name] = std::move(m);
        });
    }

    // Wait for all threads to finish naturally (via channel close propagation).
    // Do NOT use threads_.clear() which calls ~jthread() → request_stop() prematurely.
    for (auto& t : threads_) {
        if (t.joinable()) t.join();
    }
    threads_.clear();

    auto t1 = std::chrono::steady_clock::now();
    wall_time_ = std::chrono::duration<double>(t1 - t0).count();

    return metrics;
}

void Scheduler::stop() noexcept {
    for (auto& t : threads_) {
        t.request_stop();
    }
    graph_.close_all_channels();
}

void Scheduler::print_metrics(
    const std::unordered_map<std::string, NodeMetrics>& metrics,
    double wall_time) {
    std::cout << std::left
              << std::setw(24) << "Node"
              << std::setw(12) << "Items"
              << std::setw(12) << "Time(s)"
              << std::setw(16) << "Throughput"
              << "Error" << '\n';
    std::cout << std::string(76, '-') << '\n';

    for (const auto& [name, m] : metrics) {
        std::cout << std::left
                  << std::setw(24) << m.name
                  << std::setw(12) << m.items_processed
                  << std::setw(12) << std::fixed << std::setprecision(4)
                  << m.elapsed_s
                  << std::setw(16) << std::fixed << std::setprecision(1)
                  << m.throughput;
        if (!m.error.empty()) std::cout << m.error;
        std::cout << '\n';
    }
    std::cout << std::string(76, '-') << '\n';
    std::cout << "Wall time: " << std::fixed << std::setprecision(4)
              << wall_time << "s\n";
}

} // namespace graphos
